From 43a86cbcbfd38e992d6f5d5009eea60f436711e4 Mon Sep 17 00:00:00 2001
From: "H. Peter Anvin" <hpa@zytor.com>
Date: Tue, 8 Sep 2026 17:33:47 -0700
Subject: [PATCH] tftpd: close descriptors on access validation errors

Close the file descriptor before returning access errors after a file has
been opened.  This avoids leaking a descriptor for rejected read and write
requests, including failed post-open truncation.

Patch from Fedora SRPM.

Signed-off-by: H. Peter Anvin <hpa@zytor.com>
---
 tftpd/tftpd.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/tftpd/tftpd.c b/tftpd/tftpd.c
index 32e4150..8287161 100644
--- a/tftpd/tftpd.c
+++ b/tftpd/tftpd.c
@@ -1701,6 +1701,7 @@ static int validate_access(char *filename, int mode,
 
     if (mode == RRQ) {
         if (!unixperms && (stbuf.st_mode & (S_IREAD >> 6)) == 0) {
+            close(fd);
             *errmsg = "File must have global read permissions";
             return (EACCESS);
         }
@@ -1710,6 +1711,7 @@ static int validate_access(char *filename, int mode,
     } else {
         if (!unixperms) {
             if ((stbuf.st_mode & (S_IWRITE >> 6)) == 0) {
+                close(fd);
                 *errmsg = "File must have global write permissions";
                 return (EACCESS);
             }
@@ -1718,6 +1720,7 @@ static int validate_access(char *filename, int mode,
 #ifdef HAVE_FTRUNCATE
 	/* We didn't get to truncate the file at open() time */
 	if (ftruncate(fd, (off_t) 0)) {
+	  close(fd);
 	  *errmsg = "Cannot reset file size";
 	  return (EACCESS);
 	}
-- 
2.55.0

